Factory client hack

AntoineDao opened the pull request:
Proposing a client based on factory methods to make the code drier. Also means we can now instantiate [dataclass](https://docs.python.org/3/library/dataclasses.html) instances as a return from api calls.

The calls follow this style:
```python
from pyspeckle import SpeckleClient

client = SpeckleClient()

client.login(email='foo@bar.com', password='IeaTaVOCadoForbrEaKFasT')

projects = client.projects.list()

project = projects[0]

project.name = 'ItsAllGoneHorriblyWrong'

client.projects.update(project.id, project.dict())

client.projects.delete(project.id)
```

I can write some more extensive docs if we want to, but let's first see if this is useful and then I'll write them up proper :smile:

AntoineDao committed:
feat(client): setup resource factory based client resources

Using resource factory methods dries up the code and facilitates changes to core functionalities
(eg: switching to async calls) as it is done in one place. Adding dataclasses through pydantic to
validate objects also allows us to return class instances instead of dictionaries which means we
could add quite a bit of functionality on top here or through inheritance in other projects.

AntoineDao committed:
refactor(speckle client): make subclass of ClientBase

left all existing methods to not break compatibility but changes self.session to self.s.

BREAKING CHANGE: client is initialised with host and version (but defaults points to hestia server
so shouldn't break too many things.

AntoineDao commented:
Fixes #3 and #2

tsvilans commented:
To list streams, we need to override the `list()` method to use `path="?omit=objects"` instead of `path="/"`.
@didimitrie I'm not sure why this is needed. If it's just `/` then the server returns an error.

tsvilans committed:
Override `list()` for streams. 

This doesn't feel optimal. Also, most of the other `list()` and `get()` functions will need a `query` parameter as well. Thoughts?

tsvilans commented:
Calling `client.objects.get(id)` returns `AttributeError: 'NoneType' object has no attribute 'parse_obj'`.

This happens in 
```
  File "\speckle\base\resource.py", line 101, in _parse_response
    return self.schema.parse_obj(response)
```

I can't find where `Resource.schema` gets set or initialized...

tsvilans commented:
> I think we can rewrite the `make_request` method to accept `kwargs` as URL query parameters or pass them in the data argument (if it's a get request then dump data as query params).

Yeah, that sounds good. `kwargs` might be nice for stacking query filters and such.

tsvilans committed:
Handle null schema in _parse_response()

Check for `self.schema` and return raw response dict if `None`.

tsvilans committed:
Set schema for SpeckleObject resource

Set `self.scheme = SpeckleObject` for `objects` Resource, as in `streams`.

tsvilans commented:
Ok, I'm inclined to go ahead and merge, as now I've got it to generally work as expected, and it's not breaking anything from the older version.

The schema model is cool, it just needs to be implemented for all objects (tried it with a Mesh and it just ignores `vertices`, `faces`, etc. because they are not defined in the `SpeckleObject` schema). So I guess there will be some sort of dispatching or something.

@didimitrie and Antoine, any last thoughts or comments?

AntoineDao commented on speckle/resources/objects.py:
```diff
@@ -46,6 +46,8 @@ def __init__(self, session, basepath):
             }
         })
+           
+        self.schema = SpeckleObject
```
Good catch! Should've definitely not PR'd this at midnight :laughing: Got a bit too excited.

AntoineDao commented:
Yeah I was thinking we could add more complex data classes for speckle objects in another PR to keep this one manageable. Pydantic will help creating different schemas depending to the `type` of the object. We can also add some fancy methods to those classes once they are set up (like area from polygon objects or line length and what not).

Other than that yeah I reckon it's good to merge :+1:

botsalot commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/speckleworks/PySpeckle/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
